mlfinlab/feature_importance/cfi.py

```diff
@@ -0,0 +1,126 @@
+"""
+An Implementation of the Clustered Feature Importance algorithm described by
+Dr Marcos Lopez de Prado in 'Clustered Feature Importance (Presentation Slides)'
+(https://papers.ssrn.com/sol3/papers.cfm?abstract_id=3517595)
+"""
+
+#Imports
+import numpy as np
+import pandas as pd
+import mlfinlab as ml
+from sklearn.metrics import log_loss
+from scipy.cluster.hierarchy import linkage, fcluster
+from mlfinlab.codependence import variation_of_information_score
+
+def get_VI_matrix(df : pd.DataFrame):
+    '''
+    Get Variation of Information (VI) matrix.
+
+    :param df: (pd.DataFrame) of features
+    :return: (pd.DataFrame) of VI matrix
+    '''
+    VI = pd.DataFrame(index=df.columns)
+    features_cols = df.columns.values
+    for col0 in features_cols:
+        vi = []
+        for col1 in features_cols:
+            x = df[col0].values
+            y = df[col1].values
+            vi.append(variation_of_information_score(x,y,n_bins=None,normalize=True))
+        VI[col0] = vi
+    return VI
+
+def cluster_features(X : pd.DataFrame, C : int = None):
+    '''
+    Get Clusters containing features subsets
+
+    (param) X : (pd.DataFrame) of features
+    (param) C : (int) number of clusters to form.
+                      Must be less the total number of features.
+                      If None then it returns optimal number of clusters
+                      decided by the ONC Algorithm.
+
+    (return) : (array) of feature subsets
+    '''
+
+    vi_matrix =  get_VI_matrix(X)
+    link = linkage(vi_matrix, 'single')
+
+    if C is not None:
+        if C >= len(X.columns):
+            raise ValueError('Number of Clusters Must be less than the number of features')
+        clusters = fcluster(link, t = C, criterion='maxclust')
+        clustered_subsets = [[f for c, f in zip(clusters, X.columns) if c == ci] for ci in range(1, C+1)]
+        return clustered_subsets
+
+    else:
+        C = len(ml.clustering.get_onc_clusters(vi_matrix)[1])
```

Comment:
Regarding the comment above , in this case ONC is used correctly cause it is applied directly to dependence matrix instead of distance, great work!